From 58aed8413c6d21f7c9d53c0aebc2dcf50e6f8fb2 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Sun, 8 Apr 2018 10:48:35 +0800
Subject: [PATCH 4/5] drm: xilinx: xilinx_drm_fb: initialize properly
 drm_format_info

Since drm_format_info is constant, we use drm_format_info to initialize it
as other drm drivers in mainline to avoid the following code walking error:

Unable to handle kernel write to read-only memory at virtual address ffffff8008ab3f10
pgd = ffffffc877e59000
[ffffff8008ab3f10] *pgd=0000000877f99003, *pud=0000000877f99003, *pmd=0000000000000000
Internal error: Oops: 9600004f [#1] PREEMPT SMP
Modules linked in: mali(O) openvswitch nf_defrag_ipv6 nf_conntrack_ipv4 nf_defrag_ipv4 nf_nat_ipv4 nf_nat nf_conntrack libcrc32c
CPU: 2 PID: 700 Comm: Xorg Tainted: G           O    4.12.21-yocto-standard #1
Hardware name: ZynqMP ZCU102 Rev1.0 (DT)
         task: ffffffc8799eb200 task.stack: ffffffc876424000
PC is at xilinx_drm_fb_create+0x118/0x1f0
LR is at xilinx_drm_fb_create+0x110/0x1f0
Starting Hostname Service...
pc : [<ffffff8008646650>] lr : [<ffffff8008646648>] pstate: 60000145
sp : ffffffc876427b40
x29: ffffffc876427b40 x28: 000000000017f800
x27: ffffffc877c96800 x26: ffffffc876427bc0
x25: ffffffc877eed200 x24: ffffffc876427cac
x23: 0000000000000001 x22: 0000000000000400
x21: ffffffc876427c98 x20: ffffffc879262b00
x19: 0000000000000001 x18: 00000000000000be
x17: 0000007f7cd1fe10 x16: ffffff800822f010
x15: 0000007f7cffbb00 x14: 0400000000000001
x13: 0000000000000300 x12: 0000000000000400
x11: 00000000ffffffff x10: 0000000000000001
x9 : 0000000000000040 x8 : 0000000000000000
x7 : ffffffc879510920 x6 : ffffffc879262b18
x5 : 0000000000000000 x4 : 0000000000000000
x3 : 0000000000000010 x2 : ffffff8008ab3f0c
x1 : 0000000000000010 x0 : 0000000000000010
Process Xorg (pid: 700, stack limit = 0xffffffc876424000)
Stack: (0xffffffc876427b40 to 0xffffffc876428000)
Call trace:
Exception stack(0xffffffc876427970 to 0xffffffc876427aa0)
7960:                                   0000000000000001 0000008000000000
7980: ffffffc876427b40 ffffff8008646650 0000000000000000 ffffff8008dc9000
79a0: ffffffc876427af0 ffffffc8799eb758 ffffffc876427a00 ffffff8008518630
79c0: ffffffc879510920 ffffffc879af8280 ffffffc876427a20 ffffff8008518630
79e0: ffffffc879510920 ffffffc879af8280 0000000000000000 0000000000000000
7a00: ffffffc879af82b0 ffffffc876427cac 0000000000000010 0000000000000010
7a20: ffffff8008ab3f0c 0000000000000010 0000000000000000 0000000000000000
7a40: ffffffc879262b18 ffffffc879510920 0000000000000000 0000000000000040
7a60: 0000000000000001 00000000ffffffff 0000000000000400 0000000000000300
7a80: 0400000000000001 0000007f7cffbb00 ffffff800822f010 0000007f7cd1fe10
[<ffffff8008646650>] xilinx_drm_fb_create+0x118/0x1f0
[<ffffff80086381f0>] drm_internal_framebuffer_create+0x298/0x468
[<ffffff8008638400>] drm_mode_addfb2+0x40/0xc0
[<ffffff800863850c>] drm_mode_addfb+0x8c/0xa8
[<ffffff80086205b0>] drm_ioctl+0x190/0x448
[<ffffff800822e990>] do_vfs_ioctl+0xb0/0x730
[<ffffff800822f09c>] SyS_ioctl+0x8c/0xa8
[<ffffff8008083730>] el0_svc_naked+0x24/0x28
Code: 9101f3a1 97fffaf3 f9401e82 294f0fa1 (39001

Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/gpu/drm/xilinx/xilinx_drm_drv.c | 3 ++-
 drivers/gpu/drm/xilinx/xilinx_drm_fb.c  | 6 ++----
 2 files changed, 4 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/xilinx/xilinx_drm_drv.c b/drivers/gpu/drm/xilinx/xilinx_drm_drv.c
index 8c21c73..a185748 100644
--- a/drivers/gpu/drm/xilinx/xilinx_drm_drv.c
+++ b/drivers/gpu/drm/xilinx/xilinx_drm_drv.c
@@ -456,7 +456,8 @@ static int xilinx_drm_load(struct drm_device *drm, unsigned long flags)
 		if (IS_ERR(private->fb)) {
 			DRM_ERROR("failed to initialize drm fb\n");
 			private->fb = NULL;
-		}
+		} else
+			DRM_INFO("Initialize drm fb Passed\n");
 	}
 	if (!private->fb)
 		dev_info(&pdev->dev, "fbdev is not initialized\n");
diff --git a/drivers/gpu/drm/xilinx/xilinx_drm_fb.c b/drivers/gpu/drm/xilinx/xilinx_drm_fb.c
index 6af29de..598c9f0 100644
--- a/drivers/gpu/drm/xilinx/xilinx_drm_fb.c
+++ b/drivers/gpu/drm/xilinx/xilinx_drm_fb.c
@@ -444,7 +444,7 @@ struct drm_framebuffer *
 	unsigned int hsub;
 	unsigned int vsub;
 	int ret;
-	int i, bpp = 0, depth = 0;
+	int i;
 
 	hsub = drm_format_horz_chroma_subsampling(mode_cmd->pixel_format);
 	vsub = drm_format_vert_chroma_subsampling(mode_cmd->pixel_format);
@@ -480,9 +480,7 @@ struct drm_framebuffer *
 		goto err_gem_object_unreference;
 	}
 
-	xilinx_fb_get_bpp_depth(mode_cmd->pixel_format, &depth, &bpp);
-	((struct drm_format_info *)(fb->base.format))->depth = (u8)depth,
-	((struct drm_format_info *)(fb->base.format))->cpp[0] = bpp / 8;
+	fb->base.format = drm_format_info(mode_cmd->pixel_format);
 
 	if (!fb->base.format->cpp[0])
 		((struct drm_format_info *)(fb->base.format))->cpp[0] =
-- 
1.8.5.2.233.g932f7e4

